pyangbind/lib/serialise.py

```
@@ -164,54 +160,271 @@ def default(self, obj, mode="default"):
+        return six.text_type(obj)
+
+    def yangt_decimal(self, obj):
+        return six.text_type(obj)
+
+    def yangt_empty(self, obj):
+        return [None] if obj else False
+
+
+class XmlYangDataSerialiser(IETFYangDataSerialiser):
+    """
+    XML can have an empty tag, and a TypedList must be treated specially, here we mark it with a deque
+    """
+
+    def yangt_typed_list(self, obj):
+        return deque([self.default(i) for i in obj])
+
+    def yangt_empty(self, obj):
+        return None
+
+
+class pybindJSONEncoderBase(json.JSONEncoder):
+    serialiser_klass = None
+
+    def encode(self, obj):
+        return json.JSONEncoder.encode(self, self.serialiser_klass().preprocess_element(obj))
+
+    def default(self, obj):
+        return self.serialiser_klass().default(obj)
+
+
+class pybindJSONEncoder(pybindJSONEncoderBase):
+    serialiser_klass = YangDataSerialiser
+
+
+class pybindIETFJSONEncoder(pybindJSONEncoderBase):
+    serialiser_klass = IETFYangDataSerialiser
+
+    @staticmethod
+    def yname_ns_func(parent_namespace, element, yname):
+        if not element._namespace == parent_namespace:
+            # if the namespace is different, then precede with the module
+            # name as per spec.
+            return "%s:%s" % (element._defining_module, yname)
+        else:
+            return yname
+
+    @staticmethod
+    def generate_element(obj, parent_namespace=None, flt=False, with_defaults=None):
+        ietf_tree_json_func = make_generate_ietf_tree(pybindIETFJSONEncoder.yname_ns_func)
+        return ietf_tree_json_func(obj, parent_namespace=parent_namespace, flt=flt, with_defaults=with_defaults)
+
+
+class EMF(objectify.ElementMaker):
+    """
+    Override ElementMaker to do netconf specific niceties, e.g.
+     - set the namespace
+     - convert underscore to hyphens for tags
+    """
+
+    def __init__(self, namespace=None, nsmap=None):
+        assert namespace or nsmap, "Must set either namespace or nsmap"
```

Can we raise a specific error type here rather than `AssertionError` please? This means that it can be caught more cleanly